feat:Geospatial Data Type and GIS Function support for milvus

Yinwei-Yu opened the pull request:
issue: #43427
pr: #37417

This pr's main goal is merge #37417 to milvus 2.5 without conflicts.

# Main Goals

1. Create and describe collections with geospatial type
2. Insert geospatial data into the insert binlog
3. Load segments containing geospatial data into memory
4. Enable query and search can display  geospatial data
5. Support using GIS funtions like ST_EQUALS in query

# Solution

1. **Add Type**: Modify the Milvus core by adding a Geospatial type in both the C++ and Go code layers, defining the Geospatial data structure and the corresponding interfaces.
2. **Dependency Libraries**: Introduce necessary geospatial data processing libraries. In the C++ source code, use Conan package management to include the GDAL library. In the Go source code, add the go-geom library to the go.mod file.
3. **Protocol Interface**: Revise the Milvus protocol to provide mechanisms for Geospatial message serialization and deserialization.
4. **Data Pipeline**: Facilitate interaction between the client and proxy using the WKT format for geospatial data. The proxy will convert all data into WKB format for downstream processing, providing column data interfaces, segment encapsulation, segment loading, payload writing, and cache block management.
5. **Query Operators**: Implement simple display and support for filter queries. Initially, focus on filtering based on spatial relationships for a single column of geospatial literal values, providing parsing and execution for query expressions.Now only support brutal search
6. **Client Modification**: Enable the client to handle user input for geospatial data and facilitate end-to-end testing.Check the modification in pymilvus.

Yinwei-Yu committed:
Add geometry support.

Support geometry in milvus.Now can insert,query and search with gis functions.
This commit includes file changing with related to 3rdparty dependencies,cpp core and go.
Here is also an example file which shows how to use the geo type and gis functions.

Signed-off-by: Yinwei Li <yinwei.li@zilliz.com>

Yinwei-Yu commented:
/run-cpu-e2e

xiaocai2333 commented:
/run-cpu-e2e

Yinwei-Yu commented on internal/core/src/common/Chunk.h:
```diff
@@ -33,6 +33,7 @@
 
 namespace milvus {
 constexpr uint64_t MMAP_STRING_PADDING = 1;
+constexpr uint64_t MMAP_GEOMETRY_PADDING = 1;
```
Yes, geometry support mmap now. In the ChunkWriter.h, we trait GeometryChunk as StringChunk:

```c++
using GeometryChunk = StringChunk;
```

And in the ChunkWriter.cpp ,it is used both in write and finish:
```c++
void
GeometryChunkWriter::write(std::shared_ptr<arrow::RecordBatchReader> data) {
...
// use 32-bit offsets to align with StringChunk layout
    size += sizeof(uint32_t) * (row_nums_ + 1) + MMAP_GEOMETRY_PADDING;
    if (file_) {
        target_ = std::make_shared<MmapChunkTarget>(*file_, file_offset_);
    } else {
        target_ = std::make_shared<MemChunkTarget>(size);
    }
....
}


std::shared_ptr<Chunk>
GeometryChunkWriter::finish() {
    // write padding, maybe not needed anymore
    // FIXME
    char padding[MMAP_GEOMETRY_PADDING];
    target_->write(padding, MMAP_GEOMETRY_PADDING);
    auto [data, size] = target_->get();
    return std::make_shared<GeometryChunk>(row_nums_, data, size, nullable_);
}
```

Yinwei-Yu commented on internal/core/src/exec/expression/GISFunctionFilterExpr.cpp:
```diff
@@ -0,0 +1,152 @@
+            } else {
+                GEOMETRY_EXECUTE_SUB_BATCH_WITH_COMPARISON(GrowingType,
+                                                           touches);
+            }
```
No, in the last line of the macro definition, there is a line of code:
```c++
return res_vec;
```
When the macro is expanded, each case branch will be returned directly, so there will be no problem of falling through

Yinwei-Yu committed:
Modify some code according to the review

Signed-off-by: Yinwei Li <yinwei.li@zilliz.com>

xiaocai2333 commented:
/set-milestone 2.5.16

xiaocai2333 commented:
/set-milestone 2.5.17

Yinwei-Yu commented:
rerun cpp-unit-test

xiaocai2333 commented:
rerun cpp-unit-test

Yinwei-Yu committed:
feat: add ValidData field to geometry field data validation

This update introduces the ValidData field in the checkGeometryFieldData function, enhancing the validation process for geometry field data.

Signed-off-by: Yinwei Li <yinwei.li@zilliz.com>

xiaocai2333 commented on internal/core/src/common/FieldDataInterface.h:
```diff
@@ -572,6 +572,68 @@ class FieldDataStringImpl : public FieldDataImpl<std::string, true> {
+            }
+            data_[length_ + i] = geometry.value();
+            i++;
+        }
```
fill valid_data_ such as JSON_array

Yinwei-Yu committed:
fix: correct FillFieldData function to use element_count instead of array_info

This commit updates the FillFieldData function to utilize element_count for better accuracy in data handling. Additionally, it introduces a null bitmap check in the FieldDataGeometryImpl class to ensure proper handling of nullable data.

Signed-off-by: Yinwei Li <yinwei.li@zilliz.com>

xiaocai2333 commented:
/run-cpu-e2e

xiaocai2333 commented:
rerun cpp-unit-test

xiaocai2333 commented:
rerun go-sdk

Yinwei-Yu commented:
rerun go-sdk

xiaocai2333 commented:
rerun ut

xiaocai2333 committed:
Merge branch 'support_geometry' into 2.5-mergegeo-li_yinwei

xiaofan-luan commented on internal/core/src/common/FieldData.cpp:
```diff
@@ -213,6 +213,22 @@ FieldDataImpl<Type, is_type_entire_row>::FillFieldData(
+                std::dynamic_pointer_cast<arrow::BinaryArray>(array);
+            std::vector<std::string> values(element_count);
+            for (size_t index = 0; index < element_count; ++index) {
+                values[index] = geometry_array->GetString(index);
```
or we can use geometry_array->GetView() to create a string view

xiaofan-luan commented on internal/core/src/common/FieldData.cpp:
```diff
@@ -213,6 +213,22 @@ FieldDataImpl<Type, is_type_entire_row>::FillFieldData(
             return FillFieldData(values.data(), element_count);
         }
+        case DataType::GEOMETRY: {
+            auto array_info =
```
does array info used anywhere?

Yinwei-Yu replied:
No.You're right.I'll delete it

xiaofan-luan commented on internal/core/src/common/Geometry.h:
```diff
@@ -0,0 +1,176 @@
+        AssertInfo(geometry != nullptr,
+                   "failed to construct geometry from wkb data");
+        geometry_.reset(geometry);
+        to_wkb_internal();
```
if it's already from wkb, to we need to to_wkb_internal?

Yinwei-Yu commented on internal/core/src/exec/expression/GISFunctionFilterExpr.cpp:
```diff
@@ -0,0 +1,152 @@
+                GEOMETRY_EXECUTE_SUB_BATCH_WITH_COMPARISON(SealedType, equals);
+            } else {
+                GEOMETRY_EXECUTE_SUB_BATCH_WITH_COMPARISON(GrowingType, equals);
+            }
```
In the last line of the macro definition, there is a line of code:

`return res_vec;`

When the macro is expanded, each case branch will be returned directly, so there will be no problem of falling through

Yinwei-Yu commented on tests/python_client/geometry_comprehensive_test.py:
```diff
@@ -0,0 +1,327 @@
+        )
+        
+        print(f"Search results length: {len(search_results)}")
+        for i, hits in enumerate(search_results):
```
This is just a example file to.May be we need to integrate the geometry test in the e2e test structure in the future.And use assert for more accuracy result there.

xiaocai2333 commented on internal/proxy/validate_util.go:
```diff
@@ -636,6 +707,43 @@ func (v *validateUtil) checkVarCharFieldData(field *schemapb.FieldData, fieldSch
+		}
+	}
+	// replace the field data with wkb data array
+	field = &schemapb.FieldData{
```
should be 
```
*field = schemapb.FieldData{
```

xiaocai2333 committed:
Update geometry proto

Signed-off-by: Cai Zhang <cai.zhang@zilliz.com>

Yinwei-Yu committed:
Refactor Geometry class and update FillFieldData implementation

- Changed the data type for geometry values from std::string to uint8_t in FillFieldData.
- Commented out calls to to_wkb_internal() in the Geometry class to prevent unnecessary WKB conversions.
- Removed unused member variables related to WKB data and size in the Geometry class.

This refactor improves memory handling and prepares the Geometry class for future enhancements.

Signed-off-by: Yinwei Li <yinwei.li@zilliz.com>

czs007 commented on internal/core/src/common/Geometry.h:
```diff
@@ -0,0 +1,176 @@
+    //                             size_);
+    // }
+
+    ~Geometry() {
```
The destructor uses OGRGeometryFactory::destroyGeometry(geometry_.release()) to free the underlying resource. While this works, a more professional approach is to use a custom deleter for the std::unique_ptr. This makes the ownership and cleanup logic more explicit and maintainable.

```
struct OGRGeometryDeleter {
    void operator()(OGRGeometry* ptr) {
        if (ptr) {
            OGRGeometryFactory::destroyGeometry(ptr);
        }
    }
};
```

Then, you would declare your geometry_ member with this custom deleter:

```std::unique_ptr<OGRGeometry, OGRGeometryDeleter> geometry_;```

Yinwei-Yu committed:
enhance:remove unused code in Geometry.h and introduce a deleter

- Introduced OGRGeometryDeleter for automatic resource management of OGRGeometry pointers.
- Updated the Geometry class to use a unique_ptr with the custom deleter, simplifying the destructor and eliminating manual memory management.
- Removed commented-out code and unused member variables related to WKB data, enhancing code clarity.

Signed-off-by: Yinwei Li <yinwei.li@zilliz.com>

Yinwei-Yu commented:
rerun go-sdk

czs007 commented on internal/parser/planparserv2/utils.go:
```diff
@@ -156,6 +156,11 @@ func getTargetType(lDataType, rDataType schemapb.DataType) (schemapb.DataType, e
 		}
 	}
+	if typeutil.IsGeometryType(lDataType) {
+		if typeutil.IsGeometryType(rDataType) || typeutil.IsJSONType(rDataType) {
```
why support rDataType as Json?

czs007 reviewed (approved):
/lgtm

Yinwei-Yu commented:
rerun cpp-unit-test

Yinwei-Yu committed:
fix:add to_wkb_string() for unittest

Signed-off-by: Yinwei Li <yinwei.li@zilliz.com>

Yinwei-Yu commented:
rerun go-sdk

xiaocai2333 commented:
/lgtm
